### DISCUSSION TITLE
Feel free to share your feedback and post questions here

### DISCUSSION THREAD

**Discussion by SebastianMC** [SebastianMC/obsidian-custom-sort#98] [General]
Here and there (community forum, discord) I find information that that someone is having trouble enabling the custom sort or that there seems to be a conflict with some other plugin. 

Not having more detailed feedback I'm unable to help, e.g. to update the documentation with some specific cases, or create some FAQ or Q&A or fix a bug, if one is found.

**Feel free to post questions here in 'Discussion' or report bugs in issues section, if you suspect a bug**

**Comment by Bojji-git** [comment]
Hi SebastianMC; it's me again.

I need your help with the following: In a vault, I want to have all files/notes not matching given rule (in my case, files without metadata: "X") sorted "< a-z" below files/notes matching given rule (i.e., files with metada: "X").

As of now, I have the following sorting-spec frontmatter ON:

![IMG_20231012_093522.jpg](https://github.com/SebastianMC/obsidian-custom-sort/assets/146337291/d6b448a1-7fb5-4c7a-a677-1e81b827ddc1)


How to have every note matching metadata: X (as seen above) go on top and every note mismatching the same rule go after those matching ones. Using this sorting-spec, I'm having mismatching notes showing up on top of the matching notes.

Thanks in advance,

Roni

**Reply by SebastianMC** [reply]
Hello @Bojji-git again!

in your current sorting specification, the line:
```/:files```
captures (matches) all files, which don't match any other criteria. And because there are no other rules, it - by definition - matches all files. And then, within the group of these files, applies sorting by the specified metadata (`Original-YYYYMMDD`). Sorting by metadata puts items not having the specific metadata in the top or in the bottom of the group, depending on the direction (asc or desc).

To force the items (files) without the specified metadata to be collected below these with metadata, you need to do two things.
First, make the specification line more restrictive to collect (match) only items (files) which actually contain the specified metadata:
```/:files with-metadata: Original-YYYYMMDD```
This step alone will do what you want.
The second step can bring more clarity, however. And it is to add (below) yet another __catch-all-unmatched-files__ statement, with correct sorting specification:
```
/:files
  < a-z
```

The final resulting sorting specification would look like:
```yaml
sorting-spec: |
  target-folder: /*
  /folders
    < a-z
  /:files with-metadata: Original-YYYYMMDD
    > a-z by-metadata: Original-YYYYMMDD
  /:files
    < a-z
```

**Reply by Bojji-git** [reply]
Wow! Works like a charm! Many thanks, SebastianMC :)

**Reply by Kirxii** [reply]
Hello SebastianMC!
I would like to know is sorting on two metadatas or more possible?
Since I'm not sure if the syntax I'm using is actually the way to do so.
![image](https://github.com/SebastianMC/obsidian-custom-sort/assets/131458787/342ef2de-4943-4e28-857b-6a8c6d813cdb)
![image](https://github.com/SebastianMC/obsidian-custom-sort/assets/131458787/c17969e9-6887-44a5-b1b8-4830104a280e)
Depends on which one is first, the other becomes irrelevant.

If it hasn't been implemented I'd gladly wait for its release

**Reply by SebastianMC** [reply]
Hi @Kirxii 

you should tell the first group of files to collect only the ones which have the `grade` metadata, and then sort that group by the value of `grade` metadata. And them, similarly, tell the second group of files to collect only these with `index` metadata, and within that group sort by the value of `index`.

To do that your starting spec:

<img width="215" alt="2023-12-14_15-16-44" src="https://github.com/SebastianMC/obsidian-custom-sort/assets/23032356/f03033d6-cf08-4df8-b8ab-ce12804e0610">

should be extended as below:

```yaml
  target-folder: Areas/School/Physics
  /:files with-metadata: grade
    < a-z by-metadata: grade
  /:files with-metadata: index
    < a-z by-metadata: index
```

The additions should be easy to recognize and are self-explanatory.
From my perspective **it is all that you need** and in a readable form.

As a side note, you might like to sort the remaining files (not having either of `grade` or `index` metadata) in that folders explicitly, by some order, e.g. by their modification date:

```yaml
  target-folder: Areas/School/Physics
  /:files with-metadata: grade
    < a-z by-metadata: grade
  /:files with-metadata: index
    < a-z by-metadata: index
  /:files
    > modified
```

---

As a curiosity, I can add that there is some smartness related to grouping and sorting by metadata, which means that if you group items by metadata using the `with-metadata:` keyword, you don't need to repeat the metadata name in the `by-metadata:` phrase.

The full example looks like below, and depending on your perspective, it could be more or less readable :-)

```yaml
  target-folder: Areas/School/Physics
  /:files with-metadata: grade
    < a-z by-metadata:
  /:files with-metadata: index
    < a-z by-metadata:
```
 
if you want to play with more advanced configurations, which have the same effect as the simple readable ones, you can exercise:

```yaml
  target-folder: Areas/School/Physics
  /+ /:files with-metadata: index
  /+ /:files with-metadata: grade
    < a-z by-metadata: grade, < a-z by-metadata: index
```

You can read the above sorting specification as: create a group of files which contain the `index` metadata or `grade` metadata (the `/+` operator combines subsequent groups into a single group of files). And then, within this group, sort by `grade`, and the secondary sorting (for items not having `grade`) is to sort by `index`. This involves multi-level sorting specification.
And it could be a bit overwhelming :-)

To end this list of advanced examples doing the same as the simple readable specs presented in the beginning, let me give the last advanced example:

```yaml
  target-folder: Areas/School/Physics
  /:files 
    < a-z by-metadata: grade, < a-z by-metadata: index
```

It is tricky. It defines a single group of all files (`/:files`). And then, using the multi-level sorting specification, it tells to put first the ones with `grade` value, sorted alphabetically, then these with `index` metadata, also sorted alphabetically.

let me know if it works (the first readable solution) for you.

**Reply by Kirxii** [reply]
Thank you @SebastianMC!
The first 3 didn't work for me but the advanced ones work as expected!

**Comment by ryo-satsuma-art** [comment]
Hi! I was wondering if there's a way to have my files auto-sort by their file extension, in a similar way to native Windows Explorer's "group by - type" functionality? Having files grouped by their filetype makes things a lot simpler for me to see them, including sorting out where notes are in a folder that might have several other files in it at the same time, which can be difficult if the one critical .md note is sandwiched between many .png, .kra, etc files.

And thank you in advance! Considering Obsidian's ability to use the File Explorer as a... file explorer, I'm a little surprised at the lack of sorting options natively included, so I feel like you're doing some very important work here.

**Reply by SebastianMC** [reply]
Hi @ryo-satsuma-art!

The short answer is that there is no automatic grouping support in the plugin. No automatic grouping support in general, which means no support for grouping by file extension, specifically.

At the same time explicit grouping is supported and it is a core part of the plugin functionality.

You can explicitly list the groups in sorting spec. For example:

```yaml
sorting-spec: |
  /:files. ....md
  /:files. ....pdf
  /:files. ....pptx
  /+ /:files. ....png
  /+ /:files. ....PNG
  /:files. ....zip
  /:files. ....jpeg
  /:files. ....json
```

In the above example spec:
- `/:files.` prefix tells to match against full file name with extension (by default extension is ignored)
- `....md` is concatenation of `...` and `.md` -> a wildcard matching any text followed by `.md`
- `/+` prefix tells to merge the adjacent groups with that prefix

The sorting imposed by the above spec:
- *.md files go first
- then .pdf files
- then .pptx files
- then image files with extension .png or .PNG
- then go .zip files
- then .jpeg
- then .json
- and the rest follows (folders and files with different name extensions).

Then, for each of the groups, or for all of them globally you can specify the sorting order.
The above example doesn't specify any order, so the default alphabetical is used

**Reply by ryo-satsuma-art** [reply]
THANK YOU so much for this detailed response!!! This is so so so helpful!

This lead me down a bit of more reading in the documentation here https://github.com/SebastianMC/obsidian-custom-sort/blob/master/docs/manual.md
and I've been experimenting with the possibilities you've presented here. Just for fun, my current sortspec file as it exists in the root of my vault:

```
---
sorting-spec: |
    target-folder: /*
    /+ /:files. ....md
    /+ /:files. ....MD
    ---+---
    /+ /:files. ....txt
    /+ /:files. ....TXT
    ---+---
    /+ /:files. ....png
    /+ /:files. ....PNG
    ---+---
    <additional file types and such here, removed for brevity>
    ---+---
    /:files
    /folders
---
```

This is making it SO much easier to navigate my vault already and there's still so much I can do to it!

Again, thank you so so so much for this addon, it's exceptionally helpful!!!

**Reply by SebastianMC** [reply]
Hi @ryo-satsuma-art !
good to hear that it works for you in this specific scenario.
Thanks for sharing your feedback, I very appreciate that - for me it is the only way to know how the plugin is used, which features provide value for users, etc. The plugin is not collecting any usage data, and typically people come only with bug and problem reports.
Thanks again!

**Comment by ilundain** [comment]
Hey @SebastianMC

I just want to say thank you for creating and sharing your awesome plugin, and also for all your efforts and responses guiding us with its configuration.  

I finally settled on Obsidian after nearly a year of exhaustive research and testing to replace the legacy Evernote Mac client I used for well over a decade.  Alas, there one last 'missing' feature nagging at me, when your plugin saved the day! 

The ability to display in the file viewer the most recently modified files on-top but within their existing alphabetical folder structure provides a very intuitive and effective way to quickly pick target files.  It took me a few tries and reading through some of your answers in the discussions, but I was surprised how in just a few lines it works perfectly.


```
---
sorting-spec: |
  target-folder: /*
  /folders
    < a-z
  /:files
    > modified
  /--hide: sortspec.md
---
```

I'm so happy that you made this plugin :)  Thank you so much for being so kind and helpful @SebastianMC !

**Reply by SebastianMC** [reply]
Hi @ilundain ,
thank you very much for your feedback!!!

**Comment by Sidclove** [comment]
<img width="1022" alt="Screenshot 2024-11-14 at 3 57 38 AM" src="https://github.com/user-attachments/assets/7acaf87b-67f6-40be-80e0-f929cc2d6721">
Hey @SebastianMC

Can u help me within sorting my files using Custom sort plugin in obsidian?
I'm confused. I want to sort the folders within Cattle Health folder, but I'm unable to do it..
What's wrong here?

**Reply by SebastianMC** [reply]
Hi @Sidclove,

at glance, the `sortspec` file exists in the folder `Cattle Health` and then in the (red) content of the file I see specification of the target folder as the vault root folder: `  target-folder: /`. That means you are specifying the order of files and folders for the root folder of the vault, which is not your intention.
To apply the order to `Cattle Health` either change the line to `  target-folder: Castle Health`, or remove the line entirely. In that case the contextual location of the `sortspec` file will imply that the specified order is to be applied to the folder in which the file exists.

Try this first.

Then the second observation is why the content of the sortspec file is displayed in red and why the red highlight in the line with `  Canvases`. This can be either a syntax error - a result of using `tab` character instead of spaces for indentation. Or it could be simply result of the vault styles. Based solely on the screenshot it can't be confirmed.

**Reply by Sidclove** [reply]
Hey @SebastianMC 
I was using tab character instead of spaces for indentation. Is that a problem. When using space, I'm getting this instead- https://i.postimg.cc/4NNw1pBq/Screenshot-2024-11-14-at-3-21-41-PM.png

Its not working for me in both scenarios- 
<img width="1043" alt="Screenshot 2024-11-14 at 3 29 24 PM" src="https://github.com/user-attachments/assets/aff8e068-5f39-41d3-8469-9931784f1c95">

**Reply by SebastianMC** [reply]
Hi @Sidclove,

ok, now I have a full clarity.

First thing - the `tab` character is forbidden in the YAML syntax of notes frontmatter (aka Obsidian properties). This is a limitation of the standard chosen by Obsidian to keep notes metadata. This limitation has nothing to do with any plugin. You simply can't use `tab` as it will break the entire frontmatter section (the properties) and cause all its content to be ignored (including sorting specs, if any). So, please in the first step replace the tabs with spaces, two spaces are a good standard. The most convenient way to do this in Obsidian is to use the `source mode` in notes editor (via the `View` menu or status bar icon / pull-down).

Second step - I examined the exact structure of your vault and realised that the folder `Cattle Health` is a subfolder of `CATTLE`. That means the full path to the folder is `CATTLE/Cattle Health` and this value has to be used for `  target-folder:`. Please then either remove the `  target-folder: ......` line or use it with the full path of the target folder `  target-folder: CATTLE/Cattle Health`.

If the two steps don't help, please attach the `sortspec` file as plain text (file) - screenshots don't show all the details required for troubleshooting.

**Comment by Sidclove** [comment]
Hey @SebastianMC
U are a wise guy. U asked for the right thing at last. It was the problem due to not using space. Now everything works perfect. Great plugin!!!

**Comment by Chapien** [comment]
Hi @SebastianMC , is it possible to manually sort by metadata, in an explicit order? I have a reading list, with a property called 'status' that can have three values: reading, unread, complete. I want to manually sort it so that those with the reading status are on top, then unread, then complete. However, I can only see how to sort with implicit (ie, by alphabetical order) instructions. Is there any way I can list a pre-defined order of priority? Ie hard define the order as "reading, unread, complete" based on the metadata value.

**Reply by SebastianMC** [reply]
Hi @Chapien, unfortunately this kind of advanced sorting is not supported by the plugin. And at glance I don't see a quick workaround or an easy way to implement that feature.

At the same time I need to sleep on that - create #195 ticket to keep it on my list.
